File: utils/src/test/kotlin/org/jetbrains/bsp/bazel/utils/time/FormatTest.java
package org.jetbrains.bsp.bazel.commons;

import static org.assertj.core.api.Assertions.assertThat;

import java.time.Duration;
import org.junit.jupiter.api.Test;

public class FormatTest {

  @Test
  public void formatMillis() {
    test(Duration.ofMillis(100), "100ms");
  }

  @Test
  public void formatMillisPrecise() {
    test(Duration.ofMillis(153), "153ms");
  }

  @Test
  public void formatSeconds() {
    test(Duration.ofMillis(1000), "1s");
  }

  @Test
  public void formatSecondsAndMillis() {
    test(Duration.ofMillis(1500), "1.5s");
  }

  @Test
  public void formatSecondsAndMillisLowerRounding() {
    test(Duration.ofMillis(1502), "1.5s");
  }

  @Test
  public void formatSecondsAndMillisUpperRounding() {
    test(Duration.ofMillis(1552), "1.6s");
  }

  @Test
  public void formatMinutes() {
    test(Duration.ofMinutes(3), "3m");
  }

  @Test
  public void formatMinutesAndSeconds() {
    test(Duration.ofSeconds(90), "1m 30s");
  }

  @Test
  public void formatMinutesAndSecondsAndMillis() {
    test(Duration.ofMillis(90 * 1000 + 501), "1m 31s");
  }

  private void test(Duration duration, String expected) {
    assertThat(Format.duration(duration)).isEqualTo(expected);
  }
}

File: commons/src/test/java/org/jetbrains/bsp/bazel/commons/FormatTest.kt
package org.jetbrains.bsp.bazel.commons

import io.kotest.matchers.shouldBe
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import java.time.Duration

private fun getData() =
    listOf(
        //           given                        expected
        Arguments.of(Duration.ofMillis(100), "100ms"),
        Arguments.of(Duration.ofMillis(153), "153ms"),
        Arguments.of(Duration.ofMillis(1000), "1s"),
        Arguments.of(Duration.ofMillis(1500), "1.5s"),
        Arguments.of(Duration.ofMillis(1502), "1.5s"),
        Arguments.of(Duration.ofMillis(1552), "1.6s"),
        Arguments.of(Duration.ofMinutes(3), "3m"),
        Arguments.of(Duration.ofSeconds(90), "1m 30s"),
        Arguments.of(Duration.ofMillis((90 * 1000 + 501).toLong()), "1m 31s")
    )

class FormatTest {

    @ParameterizedTest(name = "Format.duration({0}) should be {1}")
    @MethodSource("org.jetbrains.bsp.bazel.commons.FormatTestKt#getData")
    fun `should format duration properly`(duration: Duration, expectedFormattedDuration: String) {
        // then
        Format.duration(duration) shouldBe expectedFormattedDuration
    }
}
